conans/client/build/compiler_id.py

```diff
@@ -0,0 +1,233 @@
+    except ValueError:
+        return UNKNOWN_COMPILER
+    except TypeError:
+        return UNKNOWN_COMPILER
+
+
+def _detect_compiler_id_internal(executable, runner=None):
+    runner = runner or ConanRunner()
+    # use a temporary file, as /dev/null might not be available on all platforms
+    tmpname = tempfile.mktemp(suffix=".c")
+    with open(tmpname, "wb") as f:
+        f.write(b"\n")
+
+    cmd = tempfile.mktemp(suffix=".cmd")
+    with open(cmd, "wb") as f:
+        f.write(b"echo off\nset MSC_CMD_FLAGS\n")
+
+    detectors = [
+        # "-dM" generate list of #define directives
+        # "-E" run only preprocessor
+        # "-x c" compiler as C code
+        # the output is of lines in form of "#define name value"
+        "-dM -E -x c",
+        "--driver-mode=g++ -dM -E -x c",  # clang-cl
+        "-c -xdumpmacros",  # SunCC,
+        # cl (Visual Studio, MSVC)
+        # "/nologo" Suppress Startup Banner
+        # "/E" Preprocess to stdout
+        # "/B1" C front-end
+        # "/c" Compile Without Linking
+        # "/TC" Specify Source File Type
+        '/nologo /E /B1 "%s" /c /TC' % cmd,
+        "/QdM /E /TC"  # icc (Intel) on Windows,
+        "-Wp,-dM -E -x c"  # QNX QCC
+    ]
+    try:
+        for detector in detectors:
+            command = '%s %s "%s"' % (executable, detector, tmpname)
+            result = StringIO()
+            if 0 == runner(command, output=result):
+                output = result.getvalue()
+                defines = dict()
+                for line in output.splitlines():
+                    tokens = line.split(' ', 3)
+                    if len(tokens) == 3 and tokens[0] == '#define':
+                        defines[tokens[1]] = tokens[2]
+                    # MSVC dumps macro definitions in single line:
+                    # "MSC_CMD_FLAGS=-D_MSC_VER=1921 -Ze"
+                    elif line.startswith("MSC_CMD_FLAGS="):
+                        line = line[len("MSC_CMD_FLAGS="):].rstrip()
+                        defines = dict()
+                        tokens = line.split()
+                        for token in tokens:
+                            if token.startswith("-D") or token.startswith("/D"):
+                                token = token[2:]
+                                if '=' in token:
+                                    name, value = token.split('=', 2)
+                                else:
+                                    name, value = token, '1'
+                                defines[name] = value
+                        break
+                compiler = _parse_compiler_version(defines)
+                if compiler == UNKNOWN_COMPILER:
+                    continue
+                return compiler
+        return UNKNOWN_COMPILER
+    finally:
+        os.unlink(tmpname)
+        os.unlink(cmd)
+
+
+_compiler_id_cache = dict()
+
+def use_cache():
+    return True
+
+def detect_compiler_id(executable, runner=None):
+    if use_cache():
```

Comment:
Will the environment affect this detection? How much can the environment change between invocations? Can `executable` be `CXX` and then we should take into account all the environment variables before using this cache?

If the problem is related to tests, then we should fix it in the test codebase, not in Conan itself. **Is this _extra time_ a problem from the POV of a user?** I don't think so, how much it takes? 1 second for each Conan run? only when running the default profile autodetection?

Reply:
do you have a suggestion how to fix it in tests? I can only imagine mocking it everywhere, but it's nightmare

Reply:
executable can't be CXX, but the value of CXX. I think we run with `shell=False` so it doesn't expand environment variables.
it's not a problem from the user's POV, only from the test suite POV
it takes less than 1 second
and it only runs for the default profile autodetection (well, not necessarily a default profile, just `conan profile new <NAME> --detect` or something else triggering auto-detection).

Reply:
I'll figure out how to deal with it in the testing codebase, here it is better to focus on the feature and to be sure that we are not breaking anything, that we are detecting exactly the same thing as we were detecting before. Better to dedicate time to the tests once we know this is something we want to (and can) merge.